### PR TITLE
[FEATURE] Add a filter class that can filter FeatureMaps on overlaps

### PR DESCRIPTION
**Pull request by jpfeuffer**
## Description

Significant speedup of FFMetaboIdent

## Checklist
- [x] Make sure that you are listed in the AUTHORS file
- [x] Add relevant changes and new features to the CHANGELOG file
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes
- [x] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

### CODE REVIEW

File: src/openms/include/OpenMS/FILTERING/DATAREDUCTION/FeatureOverlapFilter.h [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+#pragma once
+
+#include <OpenMS/KERNEL/FeatureMap.h>
+
+namespace OpenMS
+{
+  class OPENMS_DLLAPI FeatureOverlapFilter
+  {
+    public:
+    static void filter(FeatureMap& fmap);
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/FILTERING/DATAREDUCTION/FeatureOverlapFilter.h:44]: maybe add some comment how overlaps are filtered (e.g., quality score)

File: src/openms/source/FILTERING/DATAREDUCTION/FeatureOverlapFilter.cpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        const auto& bb = f->getConvexHull().getBoundingBox();
+        return quadtree::Box<float>(bb.minY(),bb.minX(),bb.maxY()-bb.minY(),bb.maxX()-bb.minX());
+    };
+
+    float minMZ = fmap.getMinMZ();
+    float maxMZ = fmap.getMaxMZ();
+    float minRT = fmap.getMinRT();
+    float maxRT = fmap.getMaxRT();
+
+    quadtree::Box<float> fullExp(minMZ-1,minRT-1,maxMZ-minMZ+2,maxRT-minRT+2);
```
Comment (timosachsenberg) [inline src/openms/source/FILTERING/DATAREDUCTION/FeatureOverlapFilter.cpp:58]: what are these +-1?

Comment (jpfeuffer) [reply src/openms/source/FILTERING/DATAREDUCTION/FeatureOverlapFilter.cpp:58]: bounding box calculations are broken and yield a box slightly bigger than what updateRanges returns..

File: src/openms/source/FILTERING/DATAREDUCTION/FeatureOverlapFilter.cpp [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    for (const auto& f : fmap)
+    {
+      if (removed_uids.count(f.getUniqueId()) == 0)
+      {
+        for (const auto& overlap : quadtree.query(getBox(&f)))
+        {
+          if (overlap != &f)
+          {
+            removed_uids.insert(overlap->getUniqueId());
+            overlap->setOverallQuality(-1.); // used to filter
```
Comment (timosachsenberg) [inline src/openms/source/FILTERING/DATAREDUCTION/FeatureOverlapFilter.cpp:73]: k maybe add some comments why this keeps the one with highest quality ...

**Comment by timosachsenberg** [comment]
nice and elegant solution using modern c++. I like it.

File: src/openms/source/FILTERING/DATAREDUCTION/FeatureOverlapFilter.cpp (hunk 2 above)
Comment (timosachsenberg) [reply src/openms/source/FILTERING/DATAREDUCTION/FeatureOverlapFilter.cpp:58]: is this something we should be concerned about?

Comment (jpfeuffer) [reply src/openms/source/FILTERING/DATAREDUCTION/FeatureOverlapFilter.cpp:58]: Looks like rounding issues. Not sure how often we depend on the fact that minMZ, maxMZ, ... has to be greater or equal to all bounding boxes generated from features' convex hull. Not so much.
But it is definitely a bug.

**Review by timosachsenberg — APPROVED** [review]

**Comment by jpfeuffer** [comment]
Wait, I will address some comments. Documentation etc.

**Comment by timosachsenberg** [comment]
I will benchmark it after the merge. I don’t expect any bottlenecks but good to have the same pc as reference.

**Comment by timosachsenberg** [comment]
With old overlap test:
```
Summary statistics:
27501 targets specified
23082 features found
0 features with multiple target annotations
4419 targets without features:
- feature_3 (m=254.078888, z=1, rt=434.431061)
- feature_6 (m=188.628693, z=1, rt=586.69574)
- feature_9 (m=128.094574, z=1, rt=380.211914)
- feature_22 (m=164.493286, z=1, rt=157.692535)
- feature_28 (m=775.22644, z=1, rt=1914.739502)
- ... (4414 more)

FeatureFinderMetaboIdent took 04:53 m (wall), 04:52 m (CPU), 1.86 s (system), 04:50 m (user); Peak Memory Usage: 1764 MB.
```

new overlap test:
```
Summary statistics:
27501 targets specified
21106 features found
0 features with multiple target annotations
6395 targets without features:
- feature_6 (m=188.628693, z=1, rt=586.69574)
- feature_9 (m=128.094574, z=1, rt=380.211914)
- feature_12 (m=102.123444, z=1, rt=292.945435)
- feature_21 (m=244.001846, z=1, rt=339.54126)
- feature_22 (m=164.493286, z=1, rt=157.692535)
- ... (6390 more)

FeatureFinderMetaboIdent took 02:37 m (wall), 02:36 m (CPU), 2.99 s (system), 02:34 m (user); Peak Memory Usage: 1759 MB.

```

**Comment by jpfeuffer** [comment]
Ah ok. We could accept another lambda/comparator for the initial sorting.

**Comment by timosachsenberg** [comment]
hmm pretty sure it is because the old code checks mass traces individually.
Thinking about if we could extend the Quadtree class with another template parameter (function object) that allows to further filter the returned overlap set.
E.g., for each query only return the elements of the overlap set that also overlap with at least one mass traces of the query.

**Comment by timosachsenberg** [comment]
Another possibility would be to add bboxes of all mass traces, perform the query on them and store the link back to the feature as some payload.

**Comment by jpfeuffer** [comment]
Yes. Both would work if you get the bboxes easily. Do you also want to check which exact peak were used?

**Comment by jpfeuffer** [comment]
Maybe the first solution is easier to implement and at least as fast. Because you know that masstraces inside one feature do not overlap. On the other hand there are mostly only one or two mass traces per features.

**Comment by jpfeuffer** [comment]
Bildschirmfoto 2023-01-24 um 15 10 23
Because you can have traces like this.

**Comment by timosachsenberg** [comment]
k I added back all annotations that were present in the old FFMID code.
To this end, I added an (optional) callback to perform some operation before an overlapping feature gets removed.
In this callback I add now the meta values (ambigious refs etc.)

**Comment by timosachsenberg** [comment]
@jpfeuffer ready for a second look

**Pull request merged**